File: app/src/test/java/org/thoughtcrime/securesms/groups/v2/GroupInviteLinkUrl_InvalidGroupLinkException_Test.java
package org.thoughtcrime.securesms.groups.v2;

import androidx.annotation.NonNull;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;

import org.junit.Test;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.storageservice.protos.groups.GroupInviteLink;
import org.thoughtcrime.securesms.util.Util;
import org.whispersystems.util.Base64UrlSafe;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.Assert.assertNull;

public final class GroupInviteLinkUrl_InvalidGroupLinkException_Test {

  @Test
  public void empty_string() throws GroupInviteLinkUrl.InvalidGroupLinkException, GroupInviteLinkUrl.UnknownGroupLinkVersionException {
    assertNull(GroupInviteLinkUrl.fromUri(""));
  }

  @Test
  public void not_a_url_string() throws GroupInviteLinkUrl.InvalidGroupLinkException, GroupInviteLinkUrl.UnknownGroupLinkVersionException {
    assertNull(GroupInviteLinkUrl.fromUri("abc"));
  }

  @Test
  public void wrong_host() throws GroupInviteLinkUrl.InvalidGroupLinkException, GroupInviteLinkUrl.UnknownGroupLinkVersionException {
    assertNull(GroupInviteLinkUrl.fromUri("https://x.signal.org/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"));
  }

  @Test
  public void wrong_scheme() throws GroupInviteLinkUrl.InvalidGroupLinkException, GroupInviteLinkUrl.UnknownGroupLinkVersionException {
    assertNull(GroupInviteLinkUrl.fromUri("http://signal.group/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"));
  }

  @Test
  public void has_path() {
    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri("https://signal.group/not_expected/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"))
                      .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                      .hasMessage("No path was expected in uri");
  }

  @Test
  public void missing_ref() {
    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri("https://signal.group/"))
                       .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                       .hasMessage("No reference was in the uri");
  }

  @Test
  public void empty_ref() {
    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri("https://signal.group/#"))
                      .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                      .hasMessage("No reference was in the uri");
  }

  @Test
  public void bad_base64() {
    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri("https://signal.group/#CAESNAogpQEzURH6BON1bCS264cmTi37Yi6HTOReXZUEHdsBIgSEPCLfiL7k4wCX;mwVi31USVY"))
                      .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                      .hasCauseExactlyInstanceOf(IOException.class);
  }

  @Test
  public void bad_protobuf() {
    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri("https://signal.group/#CAESNAogpQEzURH6BON1bCS264cmTi37Yi6HTOReXZUEHdsBIgSEPCLfiL7k4wCXmwVi31USVY"))
                      .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                      .hasCauseExactlyInstanceOf(InvalidProtocolBufferException.class);
  }

  @Test
  public void version_999_url() {
    String url = "https://signal.group/#uj4zCiDMSxlNUvF4bQ3z3fYzGyZTFbJ1xEqWbPE3uZSD8bjOrxIP8NxV-0GUz3jpxMLR1rN3";

    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri(url))
                      .isInstanceOf(GroupInviteLinkUrl.UnknownGroupLinkVersionException.class)
                      .hasMessage("Url contains no known group link content");
  }

  @Test
  public void bad_master_key_length() {
    byte[]            masterKeyBytes = Util.getSecretBytes(33);
    GroupLinkPassword password       = GroupLinkPassword.createNew();

    String encoding = createEncodedProtobuf(masterKeyBytes, password.serialize());

    String url = "https://signal.group/#" + encoding;

    assertThatThrownBy(() -> GroupInviteLinkUrl.fromUri(url))
                      .isInstanceOf(GroupInviteLinkUrl.InvalidGroupLinkException.class)
                      .hasCauseExactlyInstanceOf(InvalidInputException.class);
  }

  private static String createEncodedProtobuf(@NonNull byte[] groupMasterKey,
                                              @NonNull byte[] passwordBytes)
  {
    return Base64UrlSafe.encodeBytesWithoutPadding(GroupInviteLink.newBuilder()
                                                 .setV1Contents(GroupInviteLink.GroupInviteLinkContentsV1.newBuilder()
                                                                               .setGroupMasterKey(ByteString.copyFrom(groupMasterKey))
                                                                               .setInviteLinkPassword(ByteString.copyFrom(passwordBytes)))
                                                 .build()
                                                 .toByteArray());
  }

}

File: app/src/test/java/org/thoughtcrime/securesms/groups/v2/GroupInviteLinkUrl_InvalidGroupLinkException_Test.kt
package org.thoughtcrime.securesms.groups.v2

import assertk.assertFailure
import assertk.assertThat
import assertk.assertions.hasMessage
import assertk.assertions.isInstanceOf
import assertk.assertions.isNull
import assertk.assertions.messageContains
import assertk.assertions.rootCause
import okio.ByteString
import org.junit.Test
import org.signal.core.util.Base64.encodeUrlSafeWithoutPadding
import org.signal.libsignal.zkgroup.InvalidInputException
import org.signal.storageservice.protos.groups.GroupInviteLink
import org.thoughtcrime.securesms.groups.v2.GroupInviteLinkUrl.InvalidGroupLinkException
import org.thoughtcrime.securesms.groups.v2.GroupInviteLinkUrl.UnknownGroupLinkVersionException
import org.thoughtcrime.securesms.util.Util
import java.io.IOException

@Suppress("ClassName")
class GroupInviteLinkUrl_InvalidGroupLinkException_Test {
  @Test
  fun empty_string() {
    val uri = ""
    assertThat(GroupInviteLinkUrl.fromUri(uri)).isNull()
  }

  @Test
  fun not_a_url_string() {
    val uri = "abc"
    assertThat(GroupInviteLinkUrl.fromUri(uri)).isNull()
  }

  @Test
  fun wrong_host() {
    val uri = "https://x.signal.org/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"
    assertThat(GroupInviteLinkUrl.fromUri(uri)).isNull()
  }

  @Test
  fun wrong_scheme() {
    val uri = "http://signal.group/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"
    assertThat(GroupInviteLinkUrl.fromUri(uri)).isNull()
  }

  @Test
  fun has_path() {
    val uri = "https://signal.group/not_expected/#CjQKIAD34MKnGrBkzDztTATwjXt-9LhLLCIG9pgzvmz-NN-AEhCbwyTuxDfP2mrluK779H7o"
    assertFailure { GroupInviteLinkUrl.fromUri(uri) }
      .isInstanceOf<InvalidGroupLinkException>()
      .hasMessage("No path was expected in uri")
  }

  @Test
  fun missing_ref() {
    val uri = "https://signal.group/"
    assertFailure { GroupInviteLinkUrl.fromUri(uri) }
      .isInstanceOf<InvalidGroupLinkException>()
      .hasMessage("No reference was in the uri")
  }

  @Test
  fun empty_ref() {
    val uri = "https://signal.group/#"
    assertFailure { GroupInviteLinkUrl.fromUri(uri) }
      .isInstanceOf<InvalidGroupLinkException>()
      .hasMessage("No reference was in the uri")
  }

  @Test
  fun bad_base64() {
    val uri = "https://signal.group/#CAESNAogpQEzURH6BON1bCS264cmTi37Yi6HTOReXZUEHdsBIgSEPCLfiL7k4wCX;mwVi31USVY"
    assertFailure { GroupInviteLinkUrl.fromUri(uri) }
      .isInstanceOf<InvalidGroupLinkException>()
      .rootCause()
      .isInstanceOf<IOException>()
  }

  @Test
  fun bad_protobuf() {
    val uri = "https://signal.group/#CAESNAogpQEzURH6BON1bCS264cmTi37Yi6HTOReXZUEHdsBIgSEPCLfiL7k4wCXmwVi31USVY"
    assertFailure {
      GroupInviteLinkUrl.fromUri(uri)
    }.isInstanceOf<InvalidGroupLinkException>()
      .rootCause()
      .isInstanceOf<IllegalStateException>()
  }

  @Test
  fun version_999_url() {
    val url = "https://signal.group/#uj4zCiDMSxlNUvF4bQ3z3fYzGyZTFbJ1xEqWbPE3uZSD8bjOrxIP8NxV-0GUz3jpxMLR1rN3"
    assertFailure { GroupInviteLinkUrl.fromUri(url) }
      .isInstanceOf<UnknownGroupLinkVersionException>()
      .messageContains("Url contains no known group link content")
  }

  @Test
  fun bad_master_key_length() {
    val masterKeyBytes = Util.getSecretBytes(33)
    val password = GroupLinkPassword.createNew()

    val encoding = createEncodedProtobuf(masterKeyBytes, password.serialize())

    val url = "https://signal.group/#$encoding"

    assertFailure { GroupInviteLinkUrl.fromUri(url) }
      .isInstanceOf<InvalidGroupLinkException>()
      .rootCause()
      .isInstanceOf<InvalidInputException>()
  }

  companion object {
    private fun createEncodedProtobuf(
      groupMasterKey: ByteArray,
      passwordBytes: ByteArray
    ): String {
      return encodeUrlSafeWithoutPadding(
        GroupInviteLink.Builder()
          .v1Contents(
            GroupInviteLink.GroupInviteLinkContentsV1.Builder()
              .groupMasterKey(ByteString.of(*groupMasterKey))
              .inviteLinkPassword(ByteString.of(*passwordBytes))
              .build()
          )
          .build()
          .encode()
      )
    }
  }
}
